File: src/net/mm2d/upnp/HttpRequest.java
/*
 * Copyright(C) 2016 大前良介(OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp;

import com.sun.istack.internal.NotNull;

import java.io.IOException;
import java.net.InetAddress;
import java.net.URL;

/**
 * HTTPリクエストメッセージを表現するクラス。
 *
 * @author <a href="mailto:ryo@mm2d.net">大前良介(OHMAE Ryosuke)</a>
 */
public class HttpRequest extends HttpMessage {
    private String mMethod = "GET";
    private String mUri = "";

    /**
     * インスタンス作成。
     */
    public HttpRequest() {
        super();
    }

    @Override
    public void setStartLine(@NotNull String line) throws IllegalArgumentException {
        setRequestLine(line);
    }

    /**
     * リクエストラインを設定する。
     *
     * {@link #setStartLine(String)}のエイリアス。
     *
     * @param line リクエストライン
     * @see #setStartLine(String)
     */
    public void setRequestLine(@NotNull String line) throws IllegalArgumentException {
        final String[] params = line.split(" ");
        if (params.length < 3) {
            throw new IllegalArgumentException();
        }
        setMethod(params[0]);
        setUri(params[1]);
        setVersion(params[2]);
    }

    @Override
    @NotNull
    public String getStartLine() {
        return getMethod() + " " + getUri() + " " + getVersion();
    }

    /**
     * 送信先URLを設定する。
     *
     * @param url 接続先URL
     * @throws IOException http以外を指定した場合、URLのパースエラー
     */
    public void setUrl(@NotNull URL url) throws IOException {
        setUrl(url, false);
    }

    /**
     * 接続先URLを設定する。
     *
     * @param url 接続先URL
     * @param withHostHeader trueを指定するとURLにもとづいてHOSTヘッダの設定も行う
     * @throws IOException http以外を指定した場合、URLのパースエラー
     */
    public void setUrl(@NotNull URL url, boolean withHostHeader) throws IOException {
        if (!"http".equals(url.getProtocol())) {
            throw new IOException("unsupported protocol." + url.getProtocol());
        }
        setAddress(InetAddress.getByName(url.getHost()));
        int port = url.getPort();
        if (port < 0) {
            port = 80;
        }
        setPort(port);
        setUri(url.getFile());
        if (withHostHeader) {
            setHeader(Http.HOST, getAddressString());
        }
    }

    /**
     * リクエストメソッドを返す。
     *
     * @return リクエストメソッド
     */
    @NotNull
    public String getMethod() {
        return mMethod;
    }

    /**
     * リクエストメソッドを設定する。
     *
     * @param method リクエストメソッド
     */
    public void setMethod(@NotNull String method) {
        mMethod = method;
    }

    /**
     * URI（リクエストパス）を返す。
     *
     * @return URI
     */
    @NotNull
    public String getUri() {
        return mUri;
    }

    /**
     * URI（リクエストパス）を設定する。
     *
     * @param uri URI
     */
    public void setUri(@NotNull String uri) {
        mUri = uri;
    }
}

File: mmupnp/src/main/java/net/mm2d/upnp/HttpRequest.kt
/*
 * Copyright (c) 2019 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.upnp

import net.mm2d.upnp.internal.message.HttpMessageDelegate
import net.mm2d.upnp.internal.message.HttpMessageDelegate.StartLineDelegate
import net.mm2d.upnp.util.toAddressString
import java.io.IOException
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.SocketAddress
import java.net.URL

/**
 * HTTP request message.
 *
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
class HttpRequest internal constructor(
    private val startLineDelegate: StartLine,
    private val delegate: HttpMessageDelegate
) : HttpMessage by delegate {

    internal data class StartLine(
        var method: String = Http.GET,
        var uri: String = "",
        override var version: String = Http.DEFAULT_HTTP_VERSION
    ) : StartLineDelegate {
        override fun getStartLine(): String = "$method $uri $version"

        override fun setStartLine(startLine: String) {
            val params = startLine.split(" ", limit = 3)
            if (params.size < 3) {
                throw IllegalArgumentException()
            }
            method = params[0]
            uri = params[1]
            version = params[2]
        }
    }

    /**
     * Destination address.
     */
    var address: InetAddress? = null

    /**
     * Destination port number.
     */
    var port: Int = 0

    /**
     * Returns the request method.
     *
     * @return request method
     */
    fun getMethod(): String = startLineDelegate.method

    /**
     * Set the request method.
     *
     * @param method recuest method
     */
    fun setMethod(method: String) {
        startLineDelegate.method = method
    }

    /**
     * Return the URI (request path)
     *
     * @return URI (request path)
     */
    fun getUri(): String = startLineDelegate.uri

    /**
     * Set the URI (request path)
     *
     * Setting of path only, not setting of connection destination.
     *
     * @param uri URI (request path)
     */
    fun setUri(uri: String) {
        startLineDelegate.uri = uri
    }

    /**
     * Return address and port number combination string
     *
     * @return address and port string
     */
    @Throws(IllegalStateException::class)
    fun getAddressString(): String {
        address?.let {
            return it.toAddressString(port)
        } ?: throw IllegalStateException("address must be set")
    }

    /**
     * Set the destination URL.
     *
     * @param url            destination URL
     * @param withHostHeader true: also set the HOST header based on the URL. default is false
     * @throws IOException specify something other than http, or error occurs while the URL parse
     */
    @Throws(IOException::class)
    fun setUrl(url: URL, withHostHeader: Boolean = false) {
        if (url.protocol != "http") {
            throw IOException("unsupported protocol." + url.protocol)
        }
        address = InetAddress.getByName(url.host)
        port = when {
            url.port > 65535 ->
                throw IOException("port number is too large. port=${url.port}")
            url.port < 0 ->
                Http.DEFAULT_PORT
            else ->
                url.port
        }
        setUri(url.file)
        if (withHostHeader) {
            setHeader(Http.HOST, getAddressString())
        }
    }

    /**
     * Return destination SocketAddress
     *
     * @return destination SocketAddress
     */
    @Throws(IllegalStateException::class)
    fun getSocketAddress(): SocketAddress {
        address?.let {
            return InetSocketAddress(it, port)
        } ?: throw IllegalStateException("address must be set")
    }

    /**
     * Message as String
     */
    override fun toString(): String {
        return delegate.toString()
    }

    companion object {
        /**
         * Create a new instance.
         */
        @JvmStatic
        fun create(): HttpRequest {
            val startLine = StartLine()
            val delegate = HttpMessageDelegate(startLine)
            return HttpRequest(startLine, delegate)
        }

        /**
         * Create a new instance with the same contents as the argument.
         *
         * @param original original message
         */
        @JvmStatic
        fun copy(original: HttpRequest): HttpRequest {
            val startLine = original.startLineDelegate.copy()
            val delegate = HttpMessageDelegate(startLine, original.delegate)
            return HttpRequest(startLine, delegate).also {
                it.address = original.address
                it.port = original.port
            }
        }
    }
}

